restore NumPDBViolations info of nodes, when HTTPExtender ProcessPree…

caden2016 opened the pull request:
What type of PR is this?
/kind bug

What this PR does / why we need it:
When i want to filter some hosts that is not used for my request pod in scheduler HTTPExtender ProcessPreemption. I found the NumPDBViolations info is missing when return. This PR try to restore NumPDBViolations info of nodes, when HTTPExtender ProcessPreemption, since this info will be used in subsequent filtering steps - pickOneNodeForPreemption

Which issue(s) this PR fixes:
Fixes #

Special notes for your reviewer:
Does this PR introduce a user-facing change?
```release-note
restore NumPDBViolations info of nodes, when HTTPExtender ProcessPreemption. This info will be used in subsequent filtering steps - pickOneNodeForPreemption
```

caden2016 commented:
@ahg-g  @Huang-Wei  hello, can you help to go on

Huang-Wei reviewed (commented):
Some comments:

1. The unit test would suffice to cover `convertToNodeNameToVictims` and `convertToNodeNameToMetaVictims` only.
2. Please don't change the file permissions (I noticed it's updated  100644 → 100755 )
3. Could you reword the release note and follow the PR template the complete the PR description?
4. Don't include "Merge..." commit in the PR; otherwise the CI system won't let this PR in.

caden2016 commented:
/reopen

caden2016 commented:
/check-cla

caden2016 commented:
/retest

caden2016 commented:
/test pull-kubernetes-unit

caden2016 commented:
/reopen

caden2016 committed:
restore NumPDBViolations info of nodes, when HTTPExtender ProcessPreemption. This info will be used in subsequent filtering steps - pick OneNodeForPreemption.
add unit tests for HTTPExtender.ProcessPreemption - make sure the NumPDBViolations info of node is return as it is.

caden2016 commented:
/reopen

caden2016 commented:
/retest

caden2016 commented:
/retest

caden2016 commented:
> Some comments:
> 
> 1. The unit test would suffice to cover `convertToNodeNameToVictims` and `convertToNodeNameToMetaVictims` only.
> 2. Please don't change the file permissions (I noticed it's updated  100644 → 100755 )
> 3. Could you reword the release note and follow the PR template the complete the PR description?
> 4. Don't include "Merge..." commit in the PR; otherwise the CI system won't let this PR in.

@Huang-Wei  thanks for your remind. I have changed according to what you suggested.  The unit test changed to be TestRestoreNumPDBViolations,  to cover functions changed only, also some comments added.  As the new commit e837db6.

/auto-cc

Huang-Wei commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,85 @@ func TestIsInterested(t *testing.T) {
 	}
 }
+
+// TestRestoreNumPDBViolations test functions convertToNodeNameToMetaVictims and HTTPExtender.convertToNodeNameToVictims in HTTPExtender.ProcessPreemption
```
You can cover the changes in 2 tests.

Huang-Wei commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,85 @@ func TestIsInterested(t *testing.T) {
+// Ensure the NumPDBViolations of each node is returned from external server.
+func TestRestoreNumPDBViolations(t *testing.T) {
+	se := &HTTPExtender{}
+	nodes := []string{"node1", "node2"}
+	podsInNodeList := []*v1.Pod{
+		{ObjectMeta: metav1.ObjectMeta{Name: "pod1", UID: "uid1"}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "pod2", UID: "uid2"}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "pod3", UID: "uid3"}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "pod4", UID: "uid4"}},
+	}
```
These steps along with other nodes instantiations should be abstracted into the body of each sub-test.

Huang-Wei commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,85 @@ func TestIsInterested(t *testing.T) {
+
+	for _, tc := range []struct {
+		label            string
+		nodeCacheCapable bool
```
To cover `convertToNodeNameToMetaVictims()`, you can suppose this is always true.

Huang-Wei commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,85 @@ func TestIsInterested(t *testing.T) {
+	}
+
+	for _, tc := range []struct {
+		label            string
```
`name` is more conventional.

caden2016 commented:
@Huang-Wei Thanks for your review. I made some changeds and change the test case into 2 tests to cover the changes in my code.

dims commented:
/assign @ahg-g

ahg-g commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,146 @@ func TestIsInterested(t *testing.T) {
 }
+
+func Test_convertToNodeNameToMetaVictims(t *testing.T) {
+	type args struct {
+		nodeNameToVictims map[string]*extenderv1.Victims
+	}
```
you don't need this type, just use `nodeNameToVictims map[string]*extenderv1.Victims` instead of `args args` below.

ahg-g commented on pkg/scheduler/extender.go:
```diff
@@ -254,7 +255,8 @@ func convertToNodeNameToMetaVictims(
 	for node, victims := range nodeNameToVictims {
 		metaVictims := &extenderv1.MetaVictims{
-			Pods: []*extenderv1.MetaPod{},
+			Pods:             []*extenderv1.MetaPod{},
```
the function name sounds strange, can you pls change it to just `convertToMetaVictims`

ahg-g commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,146 @@ func TestIsInterested(t *testing.T) {
 	}
 }
+
+func Test_convertToNodeNameToMetaVictims(t *testing.T) {
```
```suggestion
func TestConvertToMetaVictims(t *testing.T) {
```

ahg-g commented on pkg/scheduler/extender.go:
```diff
@@ -217,7 +217,8 @@ func (h *HTTPExtender) convertToNodeNameToVictims(
 		}
 		victims := &extenderv1.Victims{
-			Pods: []*v1.Pod{},
+			Pods:             []*v1.Pod{},
```
pls change the function name to `convertToVictims`

ahg-g commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,146 @@ func TestIsInterested(t *testing.T) {
+	}
+}
+
+func TestHTTPExtender_convertToNodeNameToVictims(t *testing.T) {
```
```suggestion
func TestConvertToVictims(t *testing.T) {
```

ahg-g commented on pkg/scheduler/extender_test.go:
```diff
@@ -388,3 +390,146 @@ func TestIsInterested(t *testing.T) {
+}
+
+func TestHTTPExtender_convertToNodeNameToVictims(t *testing.T) {
+	type args struct {
+		nodeNameToMetaVictims map[string]*extenderv1.MetaVictims
+		nodeNames             []string
+		podsInNodeList        []*v1.Pod
+		nodeInfos             framework.NodeInfoLister
+	}
```
ditto, no need for this type, just define the parameters directly in the test case struct below.

caden2016 replied:
hi @ahg-g thanks for your review. I have changed the test fun name and args according to your suggestions.

ahg-g replied:
there are still two more comments: change `convertToNodeNameToVictims` to `convertToVictims` and `convertToNodeNameToMetaVictims` to `convertToMetaVictims`

caden2016 replied:
@ahg-g OK, I have changed

ahg-g replied:
thanks, please squash the commits

caden2016 replied:
Hi @ahg-g, I have squashed the commits into one.

caden2016 commented:
/retest

ahg-g commented:
/lgtm 
/approve
